File: source/ch/cyberduck/ui/controller/TransferPromptModel.cs
// 
// Copyright (c) 2010-2013 Yves Langisch. All rights reserved.
// http://cyberduck.ch/
// 
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// Bug fixes, suggestions and comments should be sent to:
// yves@cyberduck.ch
// 

using System;
using System.Collections.Generic;
using System.Drawing;
using System.Windows.Forms;
using Ch.Cyberduck.Core;
using Ch.Cyberduck.Ui.Winforms;
using ch.cyberduck.core;
using ch.cyberduck.core.formatter;
using ch.cyberduck.core.transfer;
using ch.cyberduck.ui.action;
using ch.cyberduck.ui.threading;
using java.util;
using org.apache.log4j;

namespace Ch.Cyberduck.Ui.Controller
{
    internal abstract class TransferPromptModel
    {
        protected static Logger log = Logger.getLogger(typeof (TransferPromptModel).FullName);
        protected readonly Transfer Transfer;

        private readonly string UNKNOWN = LocaleFactory.localizedString("Unknown");
        private readonly Cache _cache = new Cache(int.MaxValue);
        private readonly TransferPromptController _controller;

        private readonly List<Path> _roots = new List<Path>();

        /**
         * Selection status map in the prompt
         */
        private readonly IDictionary<Path, CheckState> _selected = new Dictionary<Path, CheckState>();
        private readonly Session _session;

        /**
          * Transfer status determined by filters
          */
        protected Bitmap AlertIcon = IconCache.Instance.IconForName("alert");
        private TransferAction _action;
        protected IDictionary<Path, TransferStatus> _status = new Dictionary<Path, TransferStatus>();


        protected TransferPromptModel(TransferPromptController controller, Session session, Transfer transfer)
        {
            _controller = controller;
            _session = session;
            Transfer = transfer;
            _action =
                TransferAction.forName(
                    Preferences.instance()
                               .getProperty(String.Format("queue.prompt.{0}.action.default", transfer.getType().name())));
        }

        public virtual void Add(Path p)
        {
            _roots.Add(p);
        }

        public bool CanExpand(object path)
        {
            return ((Path) path).isDirectory();
        }

        public IEnumerable<Path> ChildrenGetter(object p)
        {
            Path directory = ((Path) p);
            if (null == directory)
            {
                // Root
                if (!_cache.isCached(null))
                {
                    _cache.put(null, new AttributedList(Transfer.getRoots()));
                    Filter();
                }
            }
            else if (!_cache.isCached(directory.getReference()))
            {
                _controller.Background(new TransferPromptListAction(this, _controller, _session, directory, Transfer,
                                                                    _cache));
            }
            // Return list with filtered files included
            AttributedList list = _cache.get(null == directory ? null : directory.getReference());
            for (int i = 0; i < list.size(); i++)
            {
                yield return (Path) list.get(i);
            }
        }

        public object GetName(Path path)
        {
            return path.getName();
        }

        public object GetModified(Path path)
        {
            long modificationDate = path.attributes().getModificationDate();
            if (modificationDate != -1)
            {
                return UserDefaultsDateFormatter.ConvertJavaMillisecondsToDateTime(modificationDate);
            }
            return UNKNOWN;
        }

        public object GetSize(Path path)
        {
            TransferStatus status = GetStatus(path);
            return status.getLength();
        }

        public TransferStatus GetStatus(Path path)
        {
            if(!_status.ContainsKey(path)) {
                // Transfer filter background task has not yet finished
                return new TransferStatus();
            }
            TransferStatus status = _status[path];
            return status;
        }

        public string GetSizeAsString(object size)
        {
            return SizeFormatterFactory.get().format((long) size);
        }

        public object GetIcon(Path path)
        {
            return IconCache.Instance.IconForPath(path, IconCache.IconSize.Small);
        }

        private bool IsFiltered(Path p)
        {
            return !_status.ContainsKey(p);
        }

        public bool IsSelected(Path p)
        {
            if (_selected.ContainsKey(p))
            {
                return _selected[p] == CheckState.Checked;
            }
            return true;
        }

        public CheckState GetCheckState(Object p)
        {
            Path path = (Path) p;
            if (IsFiltered(path))
            {
                return CheckState.Unchecked;
            }
            return IsSelected(path) ? CheckState.Checked : CheckState.Unchecked;
        }

        public CheckState SetCheckState(object p, CheckState newValue)
        {
            _selected[(Path) p] = newValue;
            return newValue;
        }

        public object GetWarningImage(Path path)
        {
            TransferStatus status = GetStatus(path);
            if (path.attributes().isFile())
            {
                if (status.getLength() == 0)
                {
                    return AlertIcon;
                }
            }
            return null;
        }

        public virtual object GetCreateImage(Path path)
        {
            return null;
        }

        public virtual object GetSyncGetter(Path path)
        {
            return null;
        }

        public bool IsActive(Path path)
        {
            return _status.ContainsKey(path);
        }

        /// <summary>
        /// Change transfer action and reload list of files
        /// </summary>
        /// <param name="action">Transfer action</param>
        public void SetAction(TransferAction action)
        {
            _action = action;
            Filter();
        }

        private void Filter()
        {
            _controller.background(new FilterAction(this, _controller, _session, Transfer, _action, _cache));
        }

        private class FilterAction : WorkerBackgroundAction
        {
            public FilterAction(TransferPromptModel model, TransferPromptController controller, Session session,
                                Transfer transfer, TransferAction action, Cache cache)
                : base(
                    controller, session,
                    new InnerTransferPromptFilterWorker(model, controller, session, transfer, action, cache))
            {
            }

            private class InnerTransferPromptFilterWorker : TransferPromptFilterWorker
            {
                private readonly TransferPromptController _controller;
                private readonly TransferPromptModel _model;

                public InnerTransferPromptFilterWorker(TransferPromptModel model, TransferPromptController controller,
                                                       Session session, Transfer transfer, TransferAction action,
                                                       Cache cache) : base(session, transfer, action, cache)
                {
                    _model = model;
                    _controller = controller;
                }

                public override void cleanup(object result)
                {
                    IDictionary<Path, TransferStatus> map = Utils.ConvertFromJavaMap<Path, TransferStatus>((Map) result);
                    _model._status = map;
                    _controller.ReloadData(_model._roots);
                }
            }
        }


        private class TransferPromptListAction : WorkerBackgroundAction
        {
            public TransferPromptListAction(TransferPromptModel model, TransferPromptController controller,
                                            Session session, Path directory, Transfer transfer, Cache cache)
                : base(
                    controller, session, new InnerTransferPromptListWorker(model, session, transfer, directory, cache))
            {
            }

            private class InnerTransferPromptListWorker : TransferPromptListWorker
            {
                private readonly Cache _cache;
                private readonly Path _directory;
                private readonly TransferPromptModel _model;

                public InnerTransferPromptListWorker(TransferPromptModel model, Session session, Transfer transfer,
                                                     Path directory, Cache cache) : base(session, transfer, directory)
                {
                    _model = model;
                    _directory = directory;
                    _cache = cache;
                }

                public override void cleanup(object list)
                {
                    _cache.put(_directory.getReference(), (AttributedList) list);
                    _model.Filter();
                }
            }
        }
    }
}

File: source/ch/cyberduck/ui/cocoa/TransferPromptModel.java
package ch.cyberduck.ui.cocoa;

/*
 *  Copyright (c) 2007 David Kocher. All rights reserved.
 *  http://cyberduck.ch/
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  Bug fixes, suggestions and comments should be sent to:
 *  dkocher@cyberduck.ch
 */

import ch.cyberduck.core.AttributedList;
import ch.cyberduck.core.Cache;
import ch.cyberduck.core.NSObjectPathReference;
import ch.cyberduck.core.Path;
import ch.cyberduck.core.Preferences;
import ch.cyberduck.core.Session;
import ch.cyberduck.core.formatter.SizeFormatterFactory;
import ch.cyberduck.core.transfer.Transfer;
import ch.cyberduck.core.transfer.TransferAction;
import ch.cyberduck.core.transfer.TransferStatus;
import ch.cyberduck.ui.action.TransferPromptFilterWorker;
import ch.cyberduck.ui.action.TransferPromptListWorker;
import ch.cyberduck.ui.cocoa.application.NSCell;
import ch.cyberduck.ui.cocoa.application.NSImage;
import ch.cyberduck.ui.cocoa.application.NSOutlineView;
import ch.cyberduck.ui.cocoa.application.NSTableColumn;
import ch.cyberduck.ui.cocoa.foundation.NSAttributedString;
import ch.cyberduck.ui.cocoa.foundation.NSNumber;
import ch.cyberduck.ui.cocoa.foundation.NSObject;
import ch.cyberduck.ui.resources.IconCacheFactory;
import ch.cyberduck.ui.threading.WorkerBackgroundAction;

import org.apache.log4j.Logger;
import org.rococoa.Rococoa;
import org.rococoa.cocoa.foundation.NSInteger;

import java.util.HashMap;
import java.util.Map;

/**
 * @version $Id$
 */
public abstract class TransferPromptModel extends OutlineDataSource {
    private static Logger log = Logger.getLogger(TransferPromptModel.class);

    private TransferPromptController controller;

    private Session session;

    private Transfer transfer;

    /**
     * Selected transfer action in prompt
     */
    private TransferAction action;

    private Cache<Path> cache;

    /**
     * Selection status map in the prompt
     */
    protected Map<Path, Boolean> selected
            = new HashMap<Path, Boolean>();

    /**
     * Transfer status determined by filters
     */
    protected Map<Path, TransferStatus> status
            = new HashMap<Path, TransferStatus>();

    public enum Column {
        include,
        warning,
        filename,
        size,
    }

    /**
     * @param c        The parent window to attach the prompt
     * @param transfer Transfer
     */
    public TransferPromptModel(final TransferPromptController c, final Session session, final Transfer transfer, final Cache cache) {
        this.controller = c;
        this.session = session;
        this.transfer = transfer;
        this.cache = cache;
        this.action = TransferAction.forName(Preferences.instance().getProperty(
                String.format("queue.prompt.%s.action.default", transfer.getType().name())));
    }

    /**
     * Change transfer action and reload list of files
     *
     * @param action Transfer action
     */
    public void setAction(final TransferAction action) {
        this.action = action;
        this.filter();
    }

    public boolean isSelected(final Path file) {
        if(selected.containsKey(file)) {
            return selected.get(file);
        }
        return true;
    }

    public void setSelected(final Path file, boolean state) {
        selected.put(file, state);
    }

    /**
     * @param file File
     * @return False if transfer filter rejected file
     */
    public boolean isFiltered(final Path file) {
        return !status.containsKey(file);
    }

    protected AttributedList<Path> get(final Path directory) {
        // Return list with filtered files included
        return cache.get(null == directory ? null : directory.getReference());
    }

    public TransferStatus getStatus(final Path file) {
        if(!status.containsKey(file)) {
            // Transfer filter background task has not yet finished
            log.warn(String.format("Unknown transfer status for %s", file));
            return new TransferStatus();
        }
        return status.get(file);
    }

    protected AttributedList<Path> children(final Path directory) {
        if(null == directory) {
            // Root
            if(!cache.isCached(null)) {
                cache.put(null, new AttributedList<Path>(transfer.getRoots()));
                this.filter();
            }
        }
        else if(!cache.isCached(directory.getReference())) {
            controller.background(new WorkerBackgroundAction(controller, session,
                    new TransferPromptListWorker(session, transfer, directory) {
                        @Override
                        public void cleanup(final AttributedList<Path> list) {
                            cache.put(directory.getReference(), list);
                            filter();
                        }
                    }));
        }
        return this.get(directory);
    }

    private void filter() {
        controller.background(new WorkerBackgroundAction(controller, session,
                new TransferPromptFilterWorker(session, transfer, action, cache) {
                    @Override
                    public void cleanup(final Map<Path, TransferStatus> accepted) {
                        status = accepted;
                        controller.reloadData();
                    }
                })
        );
    }

    protected NSObject objectValueForItem(final Path file, final String identifier) {
        if(identifier.equals(Column.include.name())) {
            if(this.isFiltered(file)) {
                return NSNumber.numberWithBoolean(false);
            }
            return NSNumber.numberWithBoolean(this.isSelected(file));
        }
        if(identifier.equals(Column.filename.name())) {
            return NSAttributedString.attributedStringWithAttributes(file.getName(),
                    TableCellAttributes.browserFontLeftAlignment());
        }
        if(identifier.equals(Column.size.name())) {
            return NSAttributedString.attributedStringWithAttributes(
                    SizeFormatterFactory.get().format(this.getStatus(file).getLength()),
                    TableCellAttributes.browserFontRightAlignment());
        }
        if(identifier.equals(Column.warning.name())) {
            if(file.attributes().isFile()) {
                if(this.getStatus(file).getLength() == 0) {
                    return IconCacheFactory.<NSImage>get().iconNamed("alert.tiff");
                }
            }
            return null;
        }
        throw new IllegalArgumentException(String.format("Unknown identifier %s", identifier));
    }

    @Override
    public void outlineView_setObjectValue_forTableColumn_byItem(final NSOutlineView outlineView, final NSObject value,
                                                                 final NSTableColumn tableColumn, final NSObject item) {
        final String identifier = tableColumn.identifier();
        if(identifier.equals(Column.include.name())) {
            final Path file = cache.lookup(new NSObjectPathReference(item));
            final int state = Rococoa.cast(value, NSNumber.class).intValue();
            this.setSelected(file, state == NSCell.NSOnState);
        }
    }

    @Override
    public boolean outlineView_isItemExpandable(final NSOutlineView view, final NSObject item) {
        return cache.lookup(new NSObjectPathReference(item)).attributes().isDirectory();
    }

    @Override
    public NSInteger outlineView_numberOfChildrenOfItem(final NSOutlineView view, final NSObject item) {
        return new NSInteger(this.children(null == item ? null : cache.lookup(new NSObjectPathReference(item))).size());
    }

    @Override
    public NSObject outlineView_child_ofItem(final NSOutlineView view, final NSInteger index, final NSObject item) {
        final AttributedList<Path> children = this.get(null == item ? null : cache.lookup(new NSObjectPathReference(item)));
        return (NSObject) children.get(index.intValue()).getReference().unique();
    }

    @Override
    public NSObject outlineView_objectValueForTableColumn_byItem(final NSOutlineView view, final NSTableColumn tableColumn, final NSObject item) {
        return this.objectValueForItem(cache.lookup(new NSObjectPathReference(item)), tableColumn.identifier());
    }
}